lib/overcommit/utils/messages_utils.rb

```diff
@@ -0,0 +1,82 @@
+module Overcommit::Utils
+  # Utility to process messages
+  module MessagesUtils
+    class << self
+      # Extract file, line number, and type of message from an error/warning
+      # messages in output.
+      #
+      # Assumes each element of `output` is a separate error/warning with all
+      # information necessary to identify it.
+      #
+      # @param output_messages [Array<String>] unprocessed error/warning messages
+      # @param regex [Regexp] regular expression defining `file`, `line` and
+      #   `type` capture groups used to extract file locations and error/warning
+      #   type from each line of output
+      # @param type_categorizer [Proc] function executed against the `type`
+      #   capture group to convert it to a `:warning` or `:error` symbol. Assumes
+      #   `:error` if `nil`.
+      # @raise [Overcommit::Exceptions::MessageProcessingError] line of output did
+      #   not match regex
+      # @return [Array<Message>]
+      def extract_messages(output_messages, regex, type_categorizer = nil)
+        output_messages.map.with_index do |message, index|
+          MessageProcessor.new(message, regex, type_categorizer, index).process
+        end
+      end
+
+      MessageProcessor = Struct.new(:message, :regex, :type_categorizer, :index) do
```

Comment:
It's not clear to me why we're creating an inner class via a `Struct` that shares the same name as the [`MessageProcessor`](https://github.com/brigade/overcommit/blob/77832e556bc2113badb6079d7cd3d6e7d6e9a3aa/lib/overcommit/message_processor.rb#L10) class, even if it is isolated. Rather than create a struct and call `process` on it, why not just get rid of the struct and run:

```ruby
# Replace `MessageProcessor.new(message, regex, type_categorizer, index).process` with:
process_message(message, regex, type_categorizer, index)
```

...and redefine the `process` method below to take arguments.

This allows you to have the remaining methods in this file as regular methods. They can remain `private` and isolated from the outside world, since you're using `class << self` above.

Reply:
I do agree it is a poor choice for naming the Struct. 

The reason why I created an object to process each message within the loop is to avoid passing the context to the methods. 

```
extract_type(match, message)

extract_file(match, message)

extract_line(match, message)
```

If we move the methods to an object at individual message level we will methods without argument because now all the previous arguments are accessible via instance variables. 


```
extract_type

extract_file

extract_line
```

This way we distinguish the methods between the list level (messages) and indivual item level ( a message)

But if you still prefer to keep the methods with arguments I will change it.

Reply:
I can see the reason why using a `Struct` can save you from having to pass arguments explicitly, but I don't think this really gets you much, as it [hides the dependencies](https://medium.com/@copyconstruct/small-functions-considered-harmful-91035d316c29#1aa7) of the method (i.e. what it actually needs to operate).

Method calls with arguments passed to them give you much more information about what the method is actually doing, rather than having it read a number of instance variables internal to the object.

Reply:
Understood.